syntax = "proto3";

package RaftRpcsProto;

service RaftRpcs {
  rpc appendEntries(AppendEntriesArgs)       returns (AppendEntriesReply);
  rpc installSnapshot(InstallSnapshotArgs)  returns (InstallSnapshotReply);
  rpc requestVote(RequestVoteArgs)          returns (RequestVoteReply);
}

message LogEntry {
  string command  = 1;
  int32  logTerm  = 2;
  int32  logIndex = 3;
}

message AppendEntriesArgs {
  int32  term         = 1;
  int32  leaderId     = 2;
  int32  prevLogIndex = 3;
  int32  prevLogTerm  = 4;
  repeated LogEntry logs  = 5;
  int32  leaderCommit = 6;
}

message AppendEntriesReply {
  int32  term       = 1;
  bool   success    = 2;
  int32  missIndex  = 3;
  int32  missTerm   = 4;
}

message InstallSnapshotArgs {
  int32  term         = 1;
  int32  leaderId     = 2;
  int32  lastIncludedIndex  = 3;
  int32  lastIncludedTerm   = 4;
  bytes  data         = 5;
}

message InstallSnapshotReply {
  int32  term       = 1;
}

message RequestVoteArgs {
	int32  term         = 1;
	int32  candidateId  = 2;
	int32  lastLogIndex = 3;
	int32  lastLogTerm  = 4;
}

message RequestVoteReply {
  int32  term        = 1;
  bool   voteGranted = 2;
}